cmd/zpool/zpool_main.c

```diff
@@ -2479,61 +2498,595 @@ zpool_do_import(int argc, char **argv)
+		dash_sets = str_array_len(iostat_bottom_labels[idx]);
+
+		while (dash_sets--) {
+			if (name)
+				printf("  %*s-", column_width - 1, " ");
+			else
+				printf("  %.*s", column_width,
+				    "--------------------");
+		}
+	}
+	printf("\n");
+}
+
+
+static void
+print_iostat_separator_impl(iostat_cbdata_t *cb,
+    unsigned int force_column_width)
+{
+	print_iostat_dashes(cb, force_column_width, NULL);
+}
+
 static void
 print_iostat_separator(iostat_cbdata_t *cb)
 {
-	int i = 0;
+	print_iostat_separator_impl(cb, 0);
+}
+
+static void
+print_iostat_header_impl(iostat_cbdata_t *cb, unsigned int force_column_width,
+    const char *histo_vdev_name)
+{
+	int i, idx;
+	unsigned int text_start, rw_column_width;
+	unsigned int namewidth;
+	uint64_t flags = cb->cb_flags;
+	uint64_t f;
+	unsigned int column_width = force_column_width;
+
+	if (flags & IOS_L_HISTO_M)
+		namewidth = MAX(cb->cb_namewidth, strlen("latency"));
+	else
+		namewidth = cb->cb_namewidth;
+
+	if (flags & IOS_L_HISTO_M)
+		printf("%-*s", namewidth, histo_vdev_name);
+	else
+		printf("%*s", namewidth, "");
+
+	/* For each bit set in flags */
+	for (f = flags; f; f &= ~(1ULL << idx)) {
+		idx = ffsll(f) - 1;
+		if (!force_column_width)
+			column_width = default_column_width(cb, idx);
+
+		rw_column_width = (column_width * 2) + 2;
+
+		/* Print our top labels centered over "read  write" label. */
+		for (i = 0; iostat_top_labels[idx][i] != NULL; i++) {
+			text_start = (int) ((rw_column_width)/2 -
+			    strlen(iostat_top_labels[idx][i])/2);
+			printf("  %*s%s%*s",
+			    text_start, " ", iostat_top_labels[idx][i],
+			    (int)(rw_column_width - text_start -
+			    strlen(iostat_top_labels[idx][i])), " ");
+		}
+	}
 
-	for (i = 0; i < cb->cb_namewidth; i++)
-		(void) printf("-");
-	(void) printf("  -----  -----  -----  -----  -----  -----\n");
+	printf("\n");
+
+	printf("%-*s", namewidth, flags & IOS_L_HISTO_M ? "latency" :
+	    cb->cb_vdev_names_count ? "vdev" : "pool");
+
+	for (f = flags; f; f &= ~(1ULL << idx)) {
+		idx = ffsll(f) - 1;
+		if (!force_column_width)
+			column_width = default_column_width(cb, idx);
+
+		for (i = 0; iostat_bottom_labels[idx][i] != NULL; i++)
+			printf("  %*s", (int) column_width,
+			    iostat_bottom_labels[idx][i]);
+	}
+
+	printf("\n");
+	print_iostat_separator_impl(cb, force_column_width);
 }
 
 static void
 print_iostat_header(iostat_cbdata_t *cb)
 {
-	(void) printf("%*s     capacity     operations    bandwidth\n",
-	    cb->cb_namewidth, "");
-	(void) printf("%-*s  alloc   free   read  write   read  write\n",
-	    cb->cb_namewidth, "pool");
-	print_iostat_separator(cb);
+	print_iostat_header_impl(cb, 0, NULL);
 }
 
+
 /*
  * Display a single statistic.
  */
 static void
-print_one_stat(uint64_t value)
+print_one_stat(uint64_t value, enum zfs_nicenum_format format,
+    unsigned int column_size, boolean_t scripted)
 {
 	char buf[64];
+	zfs_nicenum_impl(value, buf, sizeof (buf), format);
```

Comment:
That'd work too, and we really wouldn't be locking ourselves in to anything since once we add `zfs_nicenum_format()` we'll need to maintain that interface.